segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
+                    return null;
+                }, executor)
+                .whenCompleteAsync((v, ex) -> readLock.unlock(), executor)
+                .thenComposeAsync(retValue -> {
+                    if (retValue != null) {
+                        return CompletableFuture.completedFuture(retValue);
+                    }
+                    // We did not find it in the buffer either.
+                    // Try to find it in store.
+                    return loadFromStore(txn, key, false)
+                            .thenApplyAsync(TransactionData::getValue, executor);
+                }, executor)
+                .whenCompleteAsync((v, ex) -> {
+                    removeFromActiveKeySet(key);
+                    GET_LATENCY.reportSuccessEvent(t.getElapsed());
+                }, executor);
+    }
+
+    private void removeFromActiveKeySet(String key) {
+        // No need to synchronize as activeKeys is already a ConcurrentHashMultiset.
+        // In case of any race with eviction logic, the key will simply be evicted next iteration.
+        // This is not incorrect and the race should be rare.
+        activeKeys.remove(key);
+    }
 
-            if (null != dataFromBuffer && null != dataFromBuffer.getValue()) {
-                // Make copy.
-                data = dataFromBuffer.toBuilder()
-                        .key(key)
-                        .value(dataFromBuffer.getValue().deepCopy())
-                        .build();
-                txnData.put(key, data);
+    private void addToActiveKeySet(String key) {
+        // No need to synchronize if the eviction is not running as activeKeys is ConcurrentHashMultiset
+        if (isEvictionRunning.get()) {
+            // However this is required when eviction is happening in background because eviction code checks the count
+            // and should evict key only if the count is zero.
+            // These two steps are not atomic hence the use of synchronized in this narrow case to prevent race.
+            synchronized (evictionLock) {
+                activeKeys.add(key);
             }
+        } else {
+            activeKeys.add(key);
         }
+    }
 
-        if (data != null) {
-            retValue = data.getValue();
+    /**
+     * Loads value from store.
+     */
+    private CompletableFuture<TransactionData> loadFromStore(MetadataTransaction txn, String key, boolean isReenterant) {
+        log.trace("Loading key from the store key = {}", key);
+        return readFromStore(key)
+                .thenApplyAsync(this::makeCopyForBuffer, executor)
+                .thenComposeAsync(copyForBuffer -> {
+                    Preconditions.checkState(null != copyForBuffer);
+                    Preconditions.checkState(null != copyForBuffer.getDbObject());
+                    if (!isReenterant) {
+                        val t = new Timer();
+                        log.debug("Acquiring write lock for {}", String.join(", ", txn.getKeysToLock()));
```

Comment:
Ouch. This `String.join` will ALWAYS be evaluated even if debug logging is off. Do not do this. Just pass in the collection and the logger will convert that to something nice.

Please fix this elsewhere if you did similar things. I see some below.

Reply:
gone